x-pack/plugins/apm/server/lib/helpers/transaction_query.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get } from 'lodash';
+import {
+  TRANSACTION_DURATION,
+  TRANSACTION_NAME
+} from '../../../common/constants';
+import { ITransactionGroup } from '../../../typings/TransactionGroup';
+
+interface ITransactionGroupSample {
+  transaction: {
+    id: string;
+    type: string;
+  };
+  context: {
+    service: {
+      name: string;
+    };
+  };
+}
+
+interface ITransactionGroupBucket {
+  key: string;
+  doc_count: number;
+  avg: {
+    value: number;
+  };
+  p95: {
+    values: {
+      '95.0': number;
+    };
+  };
+  sample: {
+    hits: {
+      hits: [
+        {
+          _source: ITransactionGroupSample;
+        }
+      ];
+    };
+  };
+}
+
+interface ITransactionGroupQueryResult {
+  aggregations: {
+    transactions: {
+      buckets?: ITransactionGroupBucket[];
+    };
+  };
+}
+
+export const TRANSACTION_GROUP_AGGREGATES = {
+  transactions: {
+    terms: {
+      field: `${TRANSACTION_NAME}.keyword`,
+      order: { avg: 'desc' },
+      size: 100
+    },
+    aggs: {
+      sample: {
+        top_hits: {
+          _source: ['context', 'transaction'],
+          size: 1,
+          sort: [{ '@timestamp': { order: 'desc' } }]
+        }
+      },
+      avg: { avg: { field: TRANSACTION_DURATION } },
+      p95: { percentiles: { field: TRANSACTION_DURATION, percents: [95] } }
+    }
+  }
+};
+
+function calculateRelativeImpacts(results: ITransactionGroup[]) {
+  const values = results.map(({ impact }) => impact);
+  const max = Math.max(...values);
+  const min = Math.min(...values);
+
+  return results.map(bucket => ({
+    ...bucket,
+    impact: ((bucket.impact - min) / (max - min)) * 100
+  }));
+}
+
+export function prepareTransactionGroups({
+  response,
+  durationMinutes
+}: {
+  response: ITransactionGroupQueryResult;
+  durationMinutes: number;
+}) {
```

Comment:
I think `prepareTransactionGroups` should just take `buckets` and not the full response. Then you can get rid of `ITransactionGroupQueryResult`.

btw. I added elasticsearch types, and it gives some pretty nice safety with few explicit types: https://github.com/sqren/kibana/blob/65a93973f823741e48894aa2875b819f03fb0a79/x-pack/plugins/apm/server/lib/traces/get_trace.ts#L62-L65

Reply:
I can do that, but I don't understand the difference?

Reply:
Oh you're saying to _replace_ the custom query result type with the ES type.

Reply:
I don't think `prepareTransactionGroups` should care about the structure of the ES response. All it cares about is the buckets. 

I think we should be careful not overtyping everything. When we make an interface typescript will blindly trust us, and I think it mostly makes sense to have interfaces when they are used multiple places. 

In this case we can get rid of `ITransactionGroupQueryResult` if it only takes the buckets.